src/compiler/checker.ts

```diff
@@ -8822,6 +8959,177 @@ module ts {
             }
         }
 
+        function checkNonThenableType(type: Type, location?: Node, message?: DiagnosticMessage) {
+            if (!allConstituentTypesHaveKind(type, TypeFlags.Any) && isTypeAssignableTo(type, getGlobalThenableType())) {

```

Comment:
Yeah, I noticed Jason had that pattern in a few places. But, as I said, there's no need to do it, you can just check the type directly because the union will have collapsed to a single type if the any type is involved.